test/loadtime/report/report.go

```diff
@@ -0,0 +1,121 @@
+package report
+
+import (
+	"math"
+	"sync"
+	"time"
+
+	"github.com/tendermint/tendermint/test/loadtime/payload"
+	"github.com/tendermint/tendermint/types"
+	"gonum.org/v1/gonum/stat"
+)
+
+// BlockStore defines the set of methods needed by the report generator from
+// Tendermint's store.Blockstore type. Using an interface allows for tests to
+// more easily simulate the required behavior without having to use the more
+// complex real API.
+type BlockStore interface {
+	Height() int64
+	Base() int64
+	LoadBlock(int64) *types.Block
+}
+
+// Report contains the data calculated from reading the timestamped transactions
+// of each block found in the blockstore.
+type Report struct {
+	Max, Min, Avg, StdDev time.Duration
+
+	// ErrorCount is the number of parsing errors encountered while reading the
+	// transaction data. Parsing errors may occur if a transaction not generated
+	// by the payload package is submitted to the chain.
+	ErrorCount int
+
+	// All contains all data points gathered from all valid transactions.
+	All []time.Duration
+}
+
+// GenerateFromBlockStore creates a Report using the data in the provided
+// BlockStore.
+func GenerateFromBlockStore(s BlockStore) (Report, error) {
+	type payloadData struct {
+		l   time.Duration
+		err error
+	}
+	type txData struct {
+		tx []byte
+		bt time.Time
+	}
+
+	const poolSize = 16
+
+	txc := make(chan txData)
+	pdc := make(chan payloadData, poolSize)
+
+	wg := &sync.WaitGroup{}
+	wg.Add(poolSize)
+	for i := 0; i < poolSize; i++ {
+		go func() {
+			defer wg.Done()
+			for b := range txc {
+				p, err := payload.FromBytes(b.tx)
```

Comment:
Parsing is slow but does not require coordination so it is done it a go routine here.

Reply:
I appreciate the effort done to parallelize transaction processing.
The (minor) problem I see is that some latencies will come out unordered. So if we get the results and want to plot them, we'll loose information.
Anyway, in that case we can set `poolSize` to 1. But would be good to be able to do it via a flag (--no-parallel, or something like that)

Reply:
> The (minor) problem I see is that some latencies will come out unordered. So if we get the results and want to plot them, we'll loose information.

This is true, but not a requirement of the tool at the moment. We can certainly change it to be ordered by adding a height / index field to the metadata given to the parallel workers and then performing a final sort of the data at the end (or heap sorting during the creation). Either sort will certainly be faster than serial parsing.

Reply:
I see. Could add a one-line comment making clear the latencies we output are unordered?